lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb

```diff
@@ -0,0 +1,162 @@
+
+          params = ['-avd', name, '-port', port]
+          params << '-no-snapshot' if cold_boot
+          params << '-wipe-data' if wipe_data
+
+          command = [@tools.emulator, *params].shelljoin + ' >/dev/null 2>/dev/null &'
+          UI.command(command)
+          system(command) # We can't use Actions.sh here because it doesn't handle `&` to run the process in the background :/
+
+          UI.message('Waiting for device to finish booting...')
+          Actions.sh(@tools.adb, '-s', serial, 'wait-for-device')
+          retry_loop(time_between_retries: BOOT_WAIT, timeout: BOOT_TIMEOUT, description: 'waiting for device to finish booting') do
+            Actions.sh(@tools.adb, '-s', serial, 'shell', 'getprop', 'sys.boot_completed').chomp == '1'
+          end
+
+          UI.success("Emulator #{name} successfully booted as `#{serial}`.")
+
+          serial
+        end
+
+        # @return [Array<Fastlane::Helper::AdbDevice>] List of currently booted emulators
+        #
+        def running_emulators
+          helper = Fastlane::Helper::AdbHelper.new(adb_path: @tools.adb)
+          helper.load_all_devices.select { |device| device.serial.include?('emulator') }
+        end
+
+        # Trigger a shutdown for all running emulators, and wait until there is no more emulator running.
+        #
+        # @param [Array<String>] serials List of emulator serials to shut down. Will shut down all of them if `nil`.
+        #
+        def shut_down_emulators!(serials: nil)
+          UI.message("Shutting down #{serials || 'all'} emulator(s)...")
+
+          emulators_list = running_emulators.map(&:serial)
+          # Get the intersection of the set of running emulators with the ones we want to shut down
+          emulators_list &= serials unless serials.nil?
+          emulators_list.each do |e|
+            Actions.sh(@tools.adb, '-s', e, 'emu', 'kill') { |_| } # ignore error if no emulator with specified serial is running
+
+            # NOTE: Alternative way of shutting down emulator would be to call the following command instead, which shuts down the emulator more gracefully:
+            # `adb -s #{e} shell reboot -p` # In case you're wondering, `-p` is for "power-off"
+            # But this alternate command:
+            #  - Requires that `-no-snapshot` was used on boot (to avoid being prompted to save current state on shutdown)
+            #  - Disconnects the emulator from `adb` (and thus disappear from `adb devices -l`) for a short amount of time,
+            #    before reconnecting to it but in an `offline` state, until `emulator` finally completely quits and it disappears
+            #    again (for good) from `adb devices --list`.
+            # This means that so if we used alternative, we couldn't really retry_loop until emulator disappears from `running_emulators` to detect
+            # that the shutdown was really complete, as we might as well accidentally detect the intermediate disconnect instead.
```

Comment:
Another area where I'd love this to be cleaner.

Especially since some old versions of the `emulator` binary end up triggering a CrashReporter alert on macOS when we send the `emu kill` command to the emulator to shut it down that way… and having a big alert on your OS every time and on every iteration of the loop that boots-then-shutdown an emulator for each locale and device that raw screenshots are taken in is… not ideal

Luckily, that CrashReporter alert I've seen in the past seems to have disappeared and not happening anymore with more recent versions of `emulator` (I'm currently using `emulator 31.3.10.0 (build_id 8807927)`), so it's probably safe to use `emu kill` after all instead of trying to be more clever with `adb shell reboot -p` 🤞